### PR TITLE
mgr/volumes: Show clone failure reason in clone status command

### PR DESCRIPTION
**Pull request by kotreshhr**
mgr/volumes: Add clone failure reason in clone status
    
Add the clone failure reason in the clone status.

The sample output is as below:
    
    $ ceph fs clone status cephfs clone_0
    {
      "status": {
        "state": "failed",
        "source": {
          "volume": "cephfs",
          "subvolume": "subvolume_0",
          "snapshot": "snapshot_0",
          "size": "52428800"
        },
        "failure": {
          "errno": "2",
          "errstr": "snapshot 'snapshot_0' does not exist"
        }
      }
    }
    
Fixes: https://tracker.ceph.com/issues/55190
Signed-off-by: Kotresh HR <khiremat@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: doc/cephfs/fs-volumes.rst [hunk 1]
```diff
@@ -340,6 +340,12 @@ A clone can be in one of the following states:
 #. `in-progress` : Clone operation is in progress
 #. `complete`    : Clone operation has successfully finished
 #. `failed`      : Clone operation has failed
+#. `canceled`    : Clone operation is cancelled by user
```
Comment (nmshelke) [inline doc/cephfs/fs-volumes.rst:343]: s/canceled/cancelled

File: doc/cephfs/fs-volumes.rst [hunk 2]
```diff
@@ -340,6 +340,12 @@ A clone can be in one of the following states:
...
+
+The failure reason of the failed clone is shown as below:
```
Comment (nmshelke) [inline doc/cephfs/fs-volumes.rst:345]: Can we reword?
```
The reason of a clone failure is shown as below:
```

File: doc/cephfs/fs-volumes.rst [hunk 3]
```diff
@@ -340,6 +340,12 @@ A clone can be in one of the following states:
...
+
+#. `errno`  : error number
+#. `errstr` : failure error string
```
Comment (nmshelke) [inline doc/cephfs/fs-volumes.rst:348]: Can say just `error string` or `error message` ?

File: doc/cephfs/fs-volumes.rst [hunk 4]
```diff
@@ -356,6 +362,28 @@ Sample output from an `in-progress` clone operation::
     }
   }
 
+.. note:: The `failure` section is shown only when clone is in `failed/canceled` state)
```
Comment (nmshelke) [inline doc/cephfs/fs-volumes.rst:365]: s/canceled/cancelled

Comment (nmshelke) [inline doc/cephfs/fs-volumes.rst:365]: Can reword:
```
.. note:: The `failure` section will be shown only if the clone is failed or cancelled.
```

File: doc/cephfs/fs-volumes.rst [hunk 5]
```diff
@@ -356,6 +362,28 @@ Sample output from an `in-progress` clone operation::
...
+
+Sample output from a `failed` clone operation::
```
Comment (nmshelke) [inline doc/cephfs/fs-volumes.rst:367]: s/from/for

File: doc/cephfs/fs-volumes.rst (hunk 1 above)
Comment (kotreshhr) [reply doc/cephfs/fs-volumes.rst:343]: Unfortunately, the typo is in actual code itself. Leaving this for now.

**Comment by kotreshhr** [comment]
Teuthology Run:
http://pulpito.front.sepia.ceph.com/khiremat-2022-04-08_08:01:09-fs:volumes-wip-khiremat-45816-clone-failure-status-1-distro-default-smithi/

**Comment by kotreshhr** [comment]
jenkins test make check

**Comment by kotreshhr** [comment]
jenkins test make check

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py [hunk 6]
```diff
@@ -794,8 +794,6 @@ def attach_snapshot(self, snapname, tgt_subvolume):
             raise VolumeException(-errno.EINVAL, "error cloning subvolume")
 
     def detach_snapshot(self, snapname, track_id):
-        if not snapname.encode('utf-8') in self.list_snapshots():
-            raise VolumeException(-errno.ENOENT, "snapshot '{0}' does not exist".format(snapname))
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:779]: Thanks for the detailed explanation in the tracker ticket.

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py [hunk 7]
```diff
@@ -139,6 +139,12 @@ def remove_clone_source(self, flush=False):
         if flush:
             self.metadata_mgr.flush()
 
+    def add_clone_failure(self, errno, error_msg):
+        self.metadata_mgr.add_section("failure")
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:143]: do we want to standardize the format of section header names? I see that section header GLOBAL is all upper case, and so is USER_METADATA.

Maybe name the section header name as CLONE_FAILURE to be explicit?

Not sure about this, but do we need class variables for class MetadataManager similar to GLOBAL_SECTION, GLOBAL_META_KEY_TYPE, etc?
So maybe in class MetadataManager,
  CLONE_FAILURE_SECTION = "CLONE_FAILURE"
  CLONE_FAILURE_META_KEY_ERRNO = "errno"
  CLONE_FAILURE_META_KEY_ERROR_MSG = "error_msg"

**Review by ajarr — COMMENTED** [review]
Overall looks good. Couple of questions

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py (hunk 7 above)
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:143]: > do we want to standardize the format of section header names? I see that section header GLOBAL is all upper case, and so is USER_METADATA.

It's not maintained in clone's `.meta` file. The `source` section name is small. So I followed the same.
> 
> Maybe name the section header name as CLONE_FAILURE to be explicit?

The command is explicitly clone status command. So I think that's the reason, it's not used in `source` section as well. I think it is fine.

> 
> Not sure about this, but do we need class variables for class MetadataManager similar to GLOBAL_SECTION, GLOBAL_META_KEY_TYPE, etc? So maybe in class MetadataManager, CLONE_FAILURE_SECTION = "CLONE_FAILURE" CLONE_FAILURE_META_KEY_ERRNO = "errno" CLONE_FAILURE_META_KEY_ERROR_MSG = "error_msg"

Not sure, it's not done for `source` section. Should we do it for all the existing ones as well then ?

Comment (ajarr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:143]: > > do we want to standardize the format of section header names? I see that section header GLOBAL is all upper case, and so is USER_METADATA.
> 
> It's not maintained in clone's `.meta` file. The `source` section name is small. So I followed the same.
> 

@kotreshhr Where does this get stored?

> > Maybe name the section header name as CLONE_FAILURE to be explicit?
> 
> The command is explicitly clone status command. So I think that's the reason, it's not used in `source` section as well. I think it is fine.
> 
> > Not sure about this, but do we need class variables for class MetadataManager similar to GLOBAL_SECTION, GLOBAL_META_KEY_TYPE, etc? So maybe in class MetadataManager, CLONE_FAILURE_SECTION = "CLONE_FAILURE" CLONE_FAILURE_META_KEY_ERRNO = "errno" CLONE_FAILURE_META_KEY_ERROR_MSG = "error_msg"
> 
> Not sure, it's not done for `source` section. Should we do it for all the existing ones as well then ?

@vshankar what do you think?

**Review by ajarr — APPROVED** [review]

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py (hunk 7 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:143]: > It's not maintained in clone's `.meta` file. The `source` section name is small. So I followed the same.
> 
> >

`add_section()` would add a new section in the config which gets written out on flush. Looks like this is adding a section in the metadata config for tracking the failure? (errno/error_msg)?

Comment (vshankar) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:143]: > Not sure about this, but do we need class variables for class MetadataManager similar to GLOBAL_SECTION, GLOBAL_META_KEY_TYPE, etc?
> So maybe in class MetadataManager,
> CLONE_FAILURE_SECTION = "CLONE_FAILURE"
> CLONE_FAILURE_META_KEY_ERRNO = "errno"
> CLONE_FAILURE_META_KEY_ERROR_MSG = "error_msg"

+1

**Review by vshankar — COMMENTED** [review]
Minor comment, otherwise looks good. Nice work.

@kotreshhr please run fs:volumes suites and share the test run.

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py (hunk 7 above)
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py]: Done

**Comment by kotreshhr** [comment]
jenkins test api

**Comment by ajarr** [comment]
looks great!

**Comment by kotreshhr** [comment]
Scheduled Teuthology Run:
http://pulpito.front.sepia.ceph.com/khiremat-2022-04-14_11:06:08-fs-wip-khiremat-45816-clone-failure-status-8-distro-default-smithi/

**Comment by kotreshhr** [comment]
> Scheduled Teuthology Run: http://pulpito.front.sepia.ceph.com/khiremat-2022-04-14_11:06:08-fs-wip-khiremat-45816-clone-failure-status-8-distro-default-smithi/

I had too kill this. Here is the new teuthology test results:
http://pulpito.front.sepia.ceph.com/khiremat-2022-04-14_16:06:02-fs:volumes-wip-khiremat-45816-clone-failure-status-8-distro-default-smithi/

**Comment by kotreshhr** [comment]
@vshankar Attached fs:volumes results. Please take a look.

**Review by vshankar — APPROVED** [review]

**Pull request merged**